stock_env.py

```diff
@@ -1,20 +1,125 @@
+                 inflation_annual=0.02,
+                 interest_annual=0.03,
+                 borrow_interest_annual=0.03,
+                 transaction_penalty=0.0001):
+        """
+        Initializes a stock environment. The environment handles penalties resulting from inflation
+        and borrowing. It ensures that the episode will exit when total cash value of assets < exit_threshold.
+
+        Args:
+            is_testing (boolean)
+            initial_cash (number)
+            buy_sell_amt (number): cash amount to buy or sell
+            exit_threshold (number): minimum total cash value of assets
+            inflation_annual (number): annual inflation rate
+            interest_annual (number): annual interest rate
+            borrow_interest_annual (number): annual stock loan fee
+            transaction_penalty (number): transaction fee percentage
+        """
+        days_per_year = 261  # number of trading days per year
+        self.initial_cash = initial_cash
+        self.buy_sell_amt = buy_sell_amt
+        self.exit_threshold = exit_threshold  # minimum total cash value of assets
+        self.inflation = inflation_annual / days_per_year  # daily penalty on all assets
+        self.interest = interest_annual / days_per_year  # daily penalty on borrowed cash
+        self.borrow_interest = borrow_interest_annual / days_per_year  # daily penalty on borrowed stocks
+        self.transaction_penalty = transaction_penalty
+
+        data = get_data()  # pricing data
+        self.pricing_data = data[1] if is_testing else data[0]
+
+    def generate_episode(self, model):
+        """
+        generate an episode of experience based on the model's current policy
+        episode stops when the end of the pricing data is reached, or the agent is broke
+        the environment handles recalculation of cash on hand, cash value of stocks, and total cash value of assets
+
+        :param model: the RL agent, which contains a policy of which actions to take
+
+        :return tuple of lists (states, actions, rewards), where each list is of episode_length.
+                episode_length is not necessarily the same as num_states, because the agent might be broke and
+                terminate the episode early
+        """
+        states = []
+        actions = []
+        rewards = []
+
+        past_num = model.past_num
+        num_stocks = model.num_stocks
+
+        timestep = past_num  # we start on day number <past_num>
+        max_timestep = tf.shape(pricing_data)[1]
+        portfolio_cash = [0] * (num_stocks + 1)  # cash value of each asset
+        portfolio_cash[num_stocks] = self.initial_cash  # cash on hand
+        portfolio_shares = [0] * num_stocks  # shares of each stock owned
+        total_cash_value = self.initial_cash
+
+        # ================ GENERATION ================
+        while total_cash_value > exit_threshold:
+            if timestep > max_timestep:  # we've reached the end of pricing_data
```

Comment:
max_timestep should be a really large number right? for example 10,000 if there are 10,000 days in the whole data set. Shouldn't we stop sooner?

Reply:
if we're training on 10 years, this would be ~2600 trading days. if that is too long, we can set a maximum.

Reply:
yea we might want to generate shorter episodes, especially at the beginning of training, when the agent is acting randomly. We don't want to take a long time to generate an episode before any training happens (i.e. I think the model would train better if we use an episode, train a bit, improve the policy, then sample another improved episode to better training.